crates/oxc_parser/src/lexer/kind.rs

```diff
@@ -3,6 +3,7 @@
 use std::fmt;
 
 use oxc_ast::Atom;
+use phf::phf_map;
```

Comment:
phf is a terrible library in terms of performance. (In my benchmarking, it is easily beatable by a general purpose hash table with a custom hash function.) phf seems to be designed for large tables [1], not run-time performance.

If you want something off-the-shelf and easy, I suggest [gperf](https://www.gnu.org/software/gperf/). Unfortunately, I think gperf cannot generate Rust code. But you can have it generate C code which you wrap in Rust. Or you can add a Rust backend to gperf, or you can write a script which rewrites gperf's C output into Rust.

In quick-lint-js, switching from gperf to my custom keyword lookup function (without inline assembly) reduced jQuery parsing times by 4-5% on my Ryzen 9 5950X. Some benchmark data for just the lookup function (higher is better):
```
Rust std::collections::HashMap (default hash function):  31.6M lookups/s
Rust phf:                                                37.9M lookups/s
C++ std::unordered_map (default hash function):          33.9M lookups/s
C++ std::unordered_map (custom hash function):           45.3M lookups/s
gperf:                                                  122.2M lookups/s
mine (production):                                      (unknown; probably around 200M lookup/s)
mine (with inline assembly):                            338.1M lookups/s 😉
```
(The "Rust std::collections::HashMap" variant used unsafe code because there was a mutable global variable. If you're allergic to unsafe, expect it to be slightly slower due to `lazy_static` overhead or `Mutex` overhead or something.)

[1] [From the docs](https://crates.io/crates/phf/0.11.1): "It currently uses the CHD algorithm and can generate a 100,000 entry map in roughly .4 seconds."

Reply:
Interesting.

I just looked at your code - it's exactly what I started imagining is possible.
Super efficient hash function based on the input assumptions (voodoo), and then the key comparison is SIMD string comparison.
Neat!